spinalcordtoolbox/scripts/sct_detect_pmj.py

```diff
@@ -256,6 +271,72 @@ def ifolder2tmp(self):
+        xrange = [int(item//px) for item in xrange]
+        xshift = int(xshift//px)
+        yshift = int(yshift//py)
+        zshift = int(zshift//pz)
+        for ix in xrange:
+            # if pattern extends towards left part of the image, then crop and pad with zeros
+            if self.rl_coord + ix + 1 + xshift > nx:
+                padding_size = self.rl_coord + ix + xshift + 1 - nx
+                src = img.data[self.pa_coord - yshift:self.pa_coord + yshift + 1,
+                               self.is_coord - zshift: self.is_coord + zshift + 1,
+                               self.rl_coord + ix - xshift: self.rl_coord + ix + xshift + 1 - padding_size]
+                src = np.pad(src, ((0, 0), (0, 0), (0, padding_size)), 'constant',
+                             constant_values=0)
+            # if pattern extends towards right part of the image, then crop and pad with zeros
+            elif self.rl_coord + ix - xshift < 0:
+                padding_size = abs(ix - xshift)
+                src = img.data[self.pa_coord - yshift:self.pa_coord + yshift + 1,
+                               self.is_coord - zshift: self.is_coord + zshift + 1,
+                               self.rl_coord + ix - xshift + padding_size: self.rl_coord + ix + xshift + 1]
+                src = np.pad(src, ((0, 0), (0, 0), (padding_size, 0)), 'constant',
+                             constant_values=0)
+            else:
+                src = img.data[self.pa_coord - yshift:self.pa_coord + yshift + 1,
+                               self.is_coord - zshift: self.is_coord + zshift + 1,
+                               self.rl_coord + ix - xshift: self.rl_coord + xshift + ix + 1]
+            target = src[:, :, ::-1]  # Mirror of src (in R-L direction)
+            # convert to 1d
+            src_1d = src.ravel()
+            target_1d = target.ravel()
+            # check if src_1d contains at least one non-zero value
+            if (src_1d.size == target_1d.size) and np.any(src_1d):
+                I_corr[ind_I] = correlation(src_1d, target_1d)
+            else:
+                allzeros = 1
+            ind_I = ind_I + 1
+        if allzeros:
+            logger.warning('Data contained zero. We probably hit the edge of the image.')
+        if np.any(I_corr):
+            # if I_corr contains at least a non-zero value
+            ind_peak = [i for i in range(len(I_corr)) if I_corr[i] == max(I_corr)][0]  # index of max along x
+            logger.info('.. Peak found: x=%s (correlation = %s)', xrange[ind_peak], I_corr[ind_peak])
+            # TODO (maybe) check if correlation is high enough compared to previous R-L coord
+        else:
+            # if I_corr contains only zeros
+            logger.warning('Correlation vector only contains zeros.')
+        # Change adjust rl_coord
+        logger.info('R-L coordinate adjusted from %s to  %s)', self.rl_coord, self.rl_coord + xrange[ind_peak])
+        self.rl_coord = self.rl_coord + xrange[ind_peak]
```

Comment:
But yeah, feel free to mark this as "resolved" and merge it as-is. :slightly_smiling_face:

Reply:
I modified the function in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3468/commits/2b7336a716a4d283e8641839d1b9a9098a0290b4, it takes the image and coordinate in PIR orientation, I could maybe also change it to RPI since it is usually in that orientation?

Reply:
I can also remove the function from the `DetectPMJ `class

Reply:
> I can also remove the function from the `DetectPMJ `class

Sure! I _think_ `image.py` would be an appropriate place for it?

> it takes the image and coordinate in PIR orientation, I could maybe also change it to RPI since it is usually in that orientation?

Ahh, that's a good point... `[x, y, z]` doesn't necessarily mean `[LR, AP, IS]`, so we do have to be careful about axis order, don't we.

I think I agree that reorienting to RPI is more in line with the convention in other parts of SCT.

Reply:
I moved the function `compute_cross_corr_3d` to `image.py` and changed the orienatation to RPI in https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3468/commits/3029835cc746665f1e3d7372e508a9028a765b12, let me know if there is anything else to change 🙂

Reply:
LGTM!